docs/intro.md

```diff
@@ -0,0 +1,1014 @@
+
+!!! tip "Advanced: look at `avg()`'s implementation"
+    The RedisGears Python API `avg()` operation is implemented by the code in [GearsBuilder.py][https://github.com/RedisGears/RedisGears/blob/0.9/src/GearsBuilder.py#L93-L103] - you're encouraged to review and compare it to the above.
+
+## Local vs. Global
+Input records are determined by the worker's data and the function's reader type. While executing a distributed operation, records may need to be shuffled - a.k.a repartitioned - and moved to other workers.
+
+!!! important
+    Repartitioning impacts performance, so try avoiding it as much as possible.
+
+We've used the `collect()` operation to move all records to the originating worker, which is one repartitioning strategy. Another strategy is to distribute the records among workers by some chosen key, and as usual, we'll use examples to cover the details.
+
+We'll set our final task to be counting the number of persons per family in our database. The quickest way to get this done is probably:
+
+```python
+def fname(x):
+  ''' Extracts the family name from a person's record '''
+  return x['value']['name'].split(' ')[1]
+
+# Count family members
+GB() \
+  .countby(fname) \
+  .run('person:*')
+
+# Expected result: [
+#   {'key': 'Pibbles', 'value': 1},
+#   {'key': 'Smith', 'value': 3},
+#   {'key': 'Sanchez', 'value': 1}
+# ]
+```
+
+Do not let the apparent simplicity of the above fool you - a lot of work done by the engine (and some Pythonic wrappers) make it happen. It should be pretty obvious what's happening here though: the [`countby()`](operations.md#countby) operation returns a count for each key in its input records. The function callback argument that it accepts is an extractor for the key, so in this case `fname()` returns the person's last name.
+
+In reality, the `countby()` operation is implemented efficiently by an assortment of other steps. This is what it would look like if coded from scratch:
+
+```python
+def fname(x):
+  ''' Extracts the family name from a person's record '''
+  return x['value']['name'].split(' ')[1]
+
+def key(x):
+  ''' Extracts the key of a record '''
+  return x['key']
+
+def counter(k, a, r):
+  ''' Counts records '''
+  return (a if a else 0) + 1
+
+def summer(k, a, r):
+  ''' Sums record values '''
+  return (a if a else 0) + r['value']
+
+# Use local and global groupby operations
+GB() \
+  .localgroupby(fname, counter) \
+  .groupby(key, summer) \
+  .run('person:*')
+
+# Expected result: the same
+```
+
+We've introduced two new operations: [`localgroupby()`](operations.md#localgroupby) and [`groupby()`](operations.md#groupby). Both perform the same type of operation, that is the grouping of records but differ in regards of where they run.
+
+The first operation, `localgroupby()`, is run locally by each shard's engine. The global `groupby()` operation is run only on the originating worker and is preceded by an implicit `collect()` operation.
```

Comment:
Groupby:
```
    FlatExecutionPlan_AddBasicStep(fep, extraxtorName, extractorArg, EXTRACTKEY);
    FlatExecutionPlan_AddBasicStep(fep, stepsNames[REPARTITION], NULL, REPARTITION);
    FlatExecutionPlan_AddBasicStep(fep, accumulateName, accumulateArg, ACCUMULATE_BY_KEY);
```
So basically groupby extract->repartition->and then the groupby.
Its running on all the shards and does not contains any collect ..

Reply:
If I understand the explanation correctly what you really intended to do here is:
.localgroupby(fname, counter) \
.collect() \
.localgroupby(key, summer)

Reply:
Fixed via 079f280

Reply:
#facepalm